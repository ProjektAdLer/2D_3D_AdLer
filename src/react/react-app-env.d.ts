/// <reference types="react-scripts" />

// Images
declare module '*.jpg';
declare module '*.png';
declare module '*.env';

// 3D types
declare module '*.glb';
declare module '*.stl';

declare module '*.jpeg';
declare module '*.gif';
declare module '*.mp4';
