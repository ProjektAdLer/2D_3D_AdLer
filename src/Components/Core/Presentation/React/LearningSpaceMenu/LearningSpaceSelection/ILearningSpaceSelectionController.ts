import { ComponentID } from "src/Components/Core/Domain/Types/EntityTypes";

export default interface ILearningSpaceSelectionController {
  onLearningSpaceClicked(spaceID: ComponentID): void;
  onLearningSpaceDoubleClicked(spaceID: ComponentID): void;
}
