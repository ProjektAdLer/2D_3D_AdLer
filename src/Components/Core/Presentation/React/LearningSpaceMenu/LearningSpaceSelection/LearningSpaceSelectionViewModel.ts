import { BooleanNode } from "src/Components/Core/Application/UseCases/CalculateLearningSpaceAvailability/Parser/BooleanSyntaxTree";
import { ComponentID } from "src/Components/Core/Domain/Types/EntityTypes";
import Observable from "src/Lib/Observable";

export interface RequiredLearningSpaceData {
  id: ComponentID;
  isCompleted: boolean;
}

export interface LearningSpaceSelectionLearningSpaceData {
  id: ComponentID;
  name: string;
  requirementsSyntaxTree: BooleanNode | null;
  isAvailable: boolean;
  isCompleted: boolean;
}

export default class LearningSpaceSelectionViewModel {
  worldID: Observable<ComponentID> = new Observable<ComponentID>(-1);

  spaces: Observable<LearningSpaceSelectionLearningSpaceData[]> =
    new Observable<LearningSpaceSelectionLearningSpaceData[]>([]);

  lastSelectedSpaceID: Observable<number> = new Observable<number>(-1);
}
