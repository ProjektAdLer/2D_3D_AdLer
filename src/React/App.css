body {
  margin: 0;
  height: 100vh;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

div{
  width: 100%;
  height: 100%;
}

section{
  display: inline;
  position: fixed;
  top: 0;
  left: 0;
  width: 120px;
  height: 100vh;
  background-color: #fff;
}

button{
  width: 100px;
  margin: 10px;
  z-index: 5;
  position: fixed;
  left: 0;
  top: 0;
}

canvas {
  width: calc(100% - 100px);
  height: 100%;
  margin-left: 100px;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

.modal {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  align-items: center;
  justify-content: center;
}

.modal-content {
  width: 1000px;
  background-color: #fff;
}

.modal-header,
.modal-footer {
  padding: 10px;
  height: 60px;
}

.modal-title {
  margin: 0;
}

.modal-body {
  padding: 10px;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
}
