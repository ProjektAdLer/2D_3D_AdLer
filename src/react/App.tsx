import React from 'react';
import './App.css';
import BabylonCanvas from './components/BabylonCanvas';

function App() {
	return (
		<React.Fragment>
			<BabylonCanvas style={{ width: '100%', position: 'absolute' }} />
		</React.Fragment>
	);
}

export default App;
