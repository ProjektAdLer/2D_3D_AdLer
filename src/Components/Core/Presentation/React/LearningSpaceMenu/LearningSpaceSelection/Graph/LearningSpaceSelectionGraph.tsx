import spaceSolved from "../../../../../../../Assets/icons/check-solution.svg";
import spaceAvailable from "../../../../../../../Assets/icons/unlocked.svg";
import spaceLocked from "../../../../../../../Assets/icons/locked.svg";

import ILearningSpaceSelectionController from "../ILearningSpaceSelectionController";
import LearningSpaceSelectionViewModel, {
  LearningSpaceSelectionLearningSpaceData,
} from "../LearningSpaceSelectionViewModel";
import ReactFlow, {
  Background,
  Controls,
  Edge,
  Node,
  NodeMouseHandler,
  NodeTypes,
  useReactFlow,
} from "reactflow";
import "reactflow/dist/style.css";
import { CSSProperties, useCallback, useEffect } from "react";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";
import LearningSpaceSelectionSpaceNode, {
  LearningSpaceSelectionSpaceNodeType,
} from "./LearningSpaceSelectionSpaceNode";
import {
  BooleanAndNode,
  BooleanIDNode,
  BooleanOrNode,
} from "src/Components/Core/Application/UseCases/CalculateLearningSpaceAvailability/Parser/BooleanSyntaxTree";
import LearningSpaceSelectionRequirementNode, {
  BooleanOperatorType,
} from "./LearningSpaceSelectionRequirementNode";
import CoreDIContainer from "~DependencyInjection/CoreDIContainer";
import ILoggerPort from "src/Components/Core/Application/Ports/Interfaces/ILoggerPort";
import CORE_TYPES from "~DependencyInjection/CoreTypes";
import { LogLevelTypes } from "src/Components/Core/Domain/Types/LogLevelTypes";
import * as dagre from "@dagrejs/dagre";

export const SPACE_NODE_WIDTH = 200; // to be used in LearningSpaceSelectionSpaceNode.tsx
export const REQ_NODE_WIDTH = 64; // to be used in LearningSpaceSelectionRequirementNode.tsx
const NODE_HEIGHT = 60;
// see https://github.com/dagrejs/dagre/wiki#configuring-the-layout for more options
const dagreLayoutConfig: dagre.GraphLabel = {
  ranksep: 25,
  nodesep: 25,
};

// node types for react flow with reference to the components
const nodeTypes: NodeTypes = {
  spaceNode: LearningSpaceSelectionSpaceNode,
  requirementNode: LearningSpaceSelectionRequirementNode,
};

export default function LearningSpaceSelectionGraph(props: {
  controller: ILearningSpaceSelectionController;
  viewModel: LearningSpaceSelectionViewModel;
}) {
  const reactFlowInstance = useReactFlow();
  const [spaces] = useObservable(props.viewModel.spaces);
  const [lastSelectedSpaceID] = useObservable(props.viewModel.selectedSpaceID);
  const logger = CoreDIContainer.get<ILoggerPort>(CORE_TYPES.ILogger);

  useEffect(() => {
    if (spaces === undefined || spaces.length === 0) return;

    const setupGraph = async () => {
      logger.log(
        LogLevelTypes.INFO,
        "LearningSpaceSelectionGraph, Graph Setup",
      );

      const requirementsTrees = createRequirementTrees(spaces);
      const spaceNodes = createSpaceNodes(spaces, lastSelectedSpaceID);

      // extract node ids from spaces and requirements for layouting
      const spaceIDs = spaceNodes.map((node) => node.id);
      const requirementsNodeIDs = requirementsTrees.nodes.map(
        (node) => node.id,
      );

      // graph layout with dagre
      const layoutedNodes = layoutNodesWithDagre(
        spaceIDs,
        requirementsNodeIDs,
        requirementsTrees.edges,
      );

      // apply node positions to react flow nodes
      const nodes = applyDagreNodePositions(
        spaceNodes,
        requirementsTrees.nodes,
        layoutedNodes,
      );

      reactFlowInstance.setNodes(nodes);
      reactFlowInstance.setEdges(requirementsTrees.edges);
    };

    setupGraph();
  }, [reactFlowInstance, spaces, lastSelectedSpaceID, logger]);

  const onNodeClickCallback = useCallback<NodeMouseHandler>(
    (event: React.MouseEvent, clickedNode: Node) => {
      // only register clicks on space nodes
      if (clickedNode.type === "spaceNode") {
        props.controller.onLearningSpaceClicked(parseInt(clickedNode.id));

        // update node data to reflect the last selected node
        const nodes = reactFlowInstance.getNodes();
        nodes.forEach((node) => {
          if (node.type === "spaceNode") {
            (node as LearningSpaceSelectionSpaceNodeType).data = {
              ...(node as LearningSpaceSelectionSpaceNodeType).data,
              lastSelected: node.id === clickedNode.id,
            };
          }
        });
        reactFlowInstance.setNodes(nodes);
      }
    },
    [props.controller, reactFlowInstance],
  );

  const onNodeDoubleClickCallback = useCallback<NodeMouseHandler>(
    (event: React.MouseEvent, clickedNode: Node) => {
      console.log("onNodeDoubleClickCallback called", clickedNode);
      // only register double clicks on space nodes
      if (clickedNode.type === "spaceNode") {
        props.controller.onLearningSpaceDoubleClicked(parseInt(clickedNode.id));
      }
    },
    [props.controller],
  );

  return (
    <div className="h-[95%] w-full lg:h-full lg:w-full">
      <ReactFlow
        defaultNodes={[]}
        nodeTypes={nodeTypes}
        nodesDraggable={false}
        nodesConnectable={false}
        onNodeClick={onNodeClickCallback}
        onNodeDoubleClick={onNodeDoubleClickCallback}
        defaultEdges={[]}
        fitView={true}
        fitViewOptions={{ padding: 0.2 }}
      >
        <Background size={2} />
        <Controls showInteractive={false} />
      </ReactFlow>
    </div>
  );
}

function createSpaceNodes(
  spaces: LearningSpaceSelectionLearningSpaceData[],
  lastSelectedSpaceID: number,
): Node[] {
  const nodes = spaces.map((space) => {
    const hasInput = space.requirementsSyntaxTree !== null;

    const hasOutput = spaces.some((inputSpace) =>
      inputSpace.requirementsSyntaxTree?.containsID(space.id),
    );

    let spaceIcon: string;
    if (space.isCompleted) spaceIcon = spaceSolved;
    else if (space.isAvailable) spaceIcon = spaceAvailable;
    else spaceIcon = spaceLocked;

    const lastSelected = space.id === lastSelectedSpaceID;

    const node: LearningSpaceSelectionSpaceNodeType = {
      id: space.id.toString(),
      data: {
        icon: spaceIcon,
        label: space.name,
        input: hasInput,
        output: hasOutput,
        lastSelected: lastSelected,
      },
      type: "spaceNode",
      position: { x: 0, y: 0 },
      connectable: false,
      deletable: false,
      className: "nopan",
    };
    return node;
  });

  return nodes;
}

function createRequirementTrees(
  spaces: LearningSpaceSelectionLearningSpaceData[],
): { nodes: Node[]; edges: Edge[] } {
  return spaces.reduce(
    (accumulatedArrays, space) => {
      // skip spaces that have no requirements
      if (space.requirementsSyntaxTree === null) return accumulatedArrays;

      for (let requirementsTreeNode of space.requirementsSyntaxTree) {
        // create a new graph edge for each node in the syntax tree
        const targetString =
          requirementsTreeNode.parentNodeID === "root"
            ? space.id.toString()
            : requirementsTreeNode.parentNodeID;

        accumulatedArrays.edges.push({
          id: requirementsTreeNode.node.ID + "-" + targetString,
          source: requirementsTreeNode.node.ID,
          target: targetString,
          style: {
            stroke: "black",
          } as CSSProperties,
        } as Edge);

        // create a new graph node if the node is a boolean operator
        if (!(requirementsTreeNode.node instanceof BooleanIDNode)) {
          let operatorTypeString: BooleanOperatorType;
          if (requirementsTreeNode.node instanceof BooleanAndNode)
            operatorTypeString = "and";
          else if (requirementsTreeNode.node instanceof BooleanOrNode)
            operatorTypeString = "or";

          accumulatedArrays.nodes.push({
            id: requirementsTreeNode.node.ID,
            data: {
              operatorType: operatorTypeString!,
            },
            position: { x: 0, y: 0 },
            type: "requirementNode",
          } as Node);
        }
      }

      // return the arrays to be used in the next iteration
      return accumulatedArrays;
    },
    // empty starting arrays
    { nodes: [], edges: [] } as { nodes: Node[]; edges: Edge[] },
  );
}

// layouts the nodes with dagre and returns a list of nodes with their positions and the ids in the label field
function layoutNodesWithDagre(
  spaceNodes: string[],
  requirementsNodes: string[],
  edges: Edge[],
): dagre.Node[] {
  const graph = new dagre.graphlib.Graph();

  // set graph configuration (needs to be an emtpy object at the least for some reason)
  graph.setGraph(dagreLayoutConfig);

  // default to assigning a new object as a label for each new edge (needs to be done for some reason)
  graph.setDefaultEdgeLabel(function () {
    return {};
  });

  spaceNodes.forEach((node) => {
    graph.setNode(node, { width: SPACE_NODE_WIDTH, height: NODE_HEIGHT });
  });
  requirementsNodes.forEach((node) => {
    graph.setNode(node, { width: SPACE_NODE_WIDTH, height: NODE_HEIGHT });
  });

  edges.forEach((edge) => {
    graph.setEdge(edge.source, edge.target);
  });

  dagre.layout(graph);

  const nodes = graph.nodes().map((node) => {
    const nodeData = graph.node(node);
    return {
      label: node, // use label field to safe node id for later use
      x: nodeData.x,
      y: nodeData.y,
      width: nodeData.width,
      height: nodeData.height,
    };
  });

  return nodes;
}

function applyDagreNodePositions(
  spaceNodes: Node[],
  requirementNodes: Node[],
  dagreNodes: dagre.Node[],
): Node[] {
  const nodesWithPositions = spaceNodes.map((node) => {
    const dagreNode = dagreNodes.find((n) => n.label === node.id);
    return {
      ...node,
      position: {
        x: dagreNode!.x - SPACE_NODE_WIDTH / 2,
        y: dagreNode!.y - NODE_HEIGHT / 2,
      },
    };
  });

  nodesWithPositions.push(
    ...requirementNodes.map((node, index) => {
      const dagreNode = dagreNodes.find((n) => n.label === node.id);
      return {
        ...node,
        position: {
          x: dagreNode!.x - REQ_NODE_WIDTH / 2,
          y: dagreNode!.y - NODE_HEIGHT / 2,
        },
      };
    }),
  );

  return nodesWithPositions;
}
