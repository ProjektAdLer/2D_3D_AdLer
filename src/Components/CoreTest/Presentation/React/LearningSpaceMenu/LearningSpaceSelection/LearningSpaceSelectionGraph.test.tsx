import { fireEvent, render, waitFor, act } from "@testing-library/react";
import { mock } from "jest-mock-extended";
import React from "react";
import ILearningSpaceSelectionController from "../../../../../Core/Presentation/React/LearningSpaceMenu/LearningSpaceSelection/ILearningSpaceSelectionController";
import LearningSpaceSelectionGraph from "../../../../../Core/Presentation/React/LearningSpaceMenu/LearningSpaceSelection/Graph/LearningSpaceSelectionGraph";
import LearningSpaceSelectionViewModel from "../../../../../Core/Presentation/React/LearningSpaceMenu/LearningSpaceSelection/LearningSpaceSelectionViewModel";
import { ReactFlowProvider } from "reactflow";
import { ElkLayoutArguments, ElkNode } from "elkjs/lib/elk.bundled.js";

// mock elk to prevent async layouting
jest.mock("elkjs/lib/elk.bundled.js", () => {
  return function () {
    return {
      layout: (graph: ElkNode, args?: ElkLayoutArguments) => {
        graph.children!.forEach((child) => {
          child.x = 42;
          child.y = 42;
        });
        return Promise.resolve(graph);
      },
    };
  };
});

describe("LearningSpaceSelectionGraph", () => {
  test("should render", () => {
    const vm = new LearningSpaceSelectionViewModel();
    const controllerMock = mock<ILearningSpaceSelectionController>();

    const { container } = render(
      <ReactFlowProvider>
        <LearningSpaceSelectionGraph
          controller={controllerMock}
          viewModel={vm}
        />
      </ReactFlowProvider>
    );
  });

  test("creates a node for each space with correct data_id's", () => {
    const vm = new LearningSpaceSelectionViewModel();
    vm.spaces.Value = [
      {
        id: 1,
        name: "test",
        isAvailable: true,
        isCompleted: true,
        requiredSpaces: [],
      },
      {
        id: 2,
        name: "test",
        isAvailable: true,
        isCompleted: false,
        requiredSpaces: [],
      },
      {
        id: 3,
        name: "test",
        isAvailable: false,
        isCompleted: false,
        requiredSpaces: [],
      },
    ];
    const controllerMock = mock<ILearningSpaceSelectionController>();

    const { container } = render(
      <ReactFlowProvider>
        <LearningSpaceSelectionGraph
          controller={controllerMock}
          viewModel={vm}
        />
      </ReactFlowProvider>
    );

    const nodes = container.querySelectorAll(".react-flow__node");
    waitFor(() => {
      expect(nodes.length).toBe(3);
      expect(nodes[0].getAttribute("data-id")).toBe("1");
      expect(nodes[1].getAttribute("data-id")).toBe("2");
      expect(nodes[2].getAttribute("data-id")).toBe("3");
    });
  });

  test("creates an edge for each required space", () => {
    const vm = new LearningSpaceSelectionViewModel();
    vm.spaces.Value = [
      {
        id: 1,
        name: "test",
        isAvailable: true,
        isCompleted: true,
        requiredSpaces: [],
      },
      {
        id: 2,
        name: "test",
        isAvailable: true,
        isCompleted: true,
        requiredSpaces: [{ id: 1, isCompleted: true }],
      },
      {
        id: 3,
        name: "test",
        isAvailable: true,
        isCompleted: true,
        requiredSpaces: [
          { id: 1, isCompleted: true },
          { id: 2, isCompleted: true },
        ],
      },
    ];
    const controllerMock = mock<ILearningSpaceSelectionController>();

    const { container } = render(
      <ReactFlowProvider>
        <LearningSpaceSelectionGraph
          controller={controllerMock}
          viewModel={vm}
        />
      </ReactFlowProvider>
    );

    const edges = container.querySelectorAll(".react-flow__edge");

    waitFor(() => {
      expect(edges.length).toBe(3);
      expect(edges[0].getAttribute("data-testid")).toContain("edge-1-2");
      expect(edges[1].getAttribute("data-testid")).toContain("edge-1-3");
      expect(edges[2].getAttribute("data-testid")).toContain("edge-2-3");
    });
  });

  test.todo("sets node position calculated by elk");

  test("calls controller with space id when a node is clicked", async () => {
    const vm = new LearningSpaceSelectionViewModel();
    const spaceID = 42;
    vm.spaces.Value = [
      {
        id: spaceID,
        name: "test",
        isAvailable: true,
        isCompleted: true,
        requiredSpaces: [],
      },
    ];
    const controllerMock = mock<ILearningSpaceSelectionController>();

    const container = render(
      <ReactFlowProvider>
        <LearningSpaceSelectionGraph
          controller={controllerMock}
          viewModel={vm}
        />
      </ReactFlowProvider>
    );
    const node = await container.findByTestId("rf__node-42");

    act(() => {
      fireEvent.click(node);
    });
    expect(controllerMock.onLearningSpaceClicked).toBeCalledWith(spaceID);
  });
});
