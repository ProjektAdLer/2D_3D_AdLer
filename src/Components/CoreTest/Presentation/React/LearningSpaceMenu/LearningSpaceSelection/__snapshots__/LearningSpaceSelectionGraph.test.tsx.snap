// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LearningSpaceSelectionGraph should render 1`] = `
<div>
  <div
    class="w-full h-[95%] lg:w-full lg:h-full"
  >
    <div
      class="react-flow"
      data-testid="rf__wrapper"
      style="width: 100%; height: 100%; overflow: hidden; position: relative; z-index: 0;"
    >
      <div
        class="react-flow__renderer"
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
      >
        <div
          class="react-flow__pane"
          style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
        >
          <div
            class="react-flow__viewport react-flow__container"
            style="transform: translate(0px,0px) scale(1);"
          >
            <svg
              class="react-flow__edges react-flow__container"
              height="100"
              style="z-index: 0;"
              width="100"
            >
              <defs />
              <g />
            </svg>
            <div
              class="react-flow__edgelabel-renderer"
            />
            <div
              class="react-flow__nodes"
              style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
            />
          </div>
        </div>
      </div>
      <svg
        class="react-flow__background"
        data-testid="rf__background"
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
      >
        <pattern
          height="20"
          id="pattern-1undefined"
          patternTransform="translate(-1,-1)"
          patternUnits="userSpaceOnUse"
          width="20"
          x="0"
          y="0"
        >
          <circle
            cx="1"
            cy="1"
            fill="#91919a"
            r="1"
          />
        </pattern>
        <rect
          fill="url(#pattern-1undefined)"
          height="100%"
          width="100%"
          x="0"
          y="0"
        />
      </svg>
      <div
        class="react-flow__panel react-flow__controls bottom left"
        data-testid="rf__controls"
        style="pointer-events: all;"
      >
        <button
          aria-label="zoom in"
          class="react-flow__controls-button react-flow__controls-zoomin"
          title="zoom in"
          type="button"
        >
          <svg
            viewBox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 18.133H18.133V32h-4.266V18.133H0v-4.266h13.867V0h4.266v13.867H32z"
            />
          </svg>
        </button>
        <button
          aria-label="zoom out"
          class="react-flow__controls-button react-flow__controls-zoomout"
          title="zoom out"
          type="button"
        >
          <svg
            viewBox="0 0 32 5"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h32v4.2H0z"
            />
          </svg>
        </button>
        <button
          aria-label="fit view"
          class="react-flow__controls-button react-flow__controls-fitview"
          title="fit view"
          type="button"
        >
          <svg
            viewBox="0 0 32 30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.692 4.63c0-.53.4-.938.939-.938h5.215V0H4.708C2.13 0 0 2.054 0 4.63v5.216h3.692V4.631zM27.354 0h-5.2v3.692h5.17c.53 0 .984.4.984.939v5.215H32V4.631A4.624 4.624 0 0027.354 0zm.954 24.83c0 .532-.4.94-.939.94h-5.215v3.768h5.215c2.577 0 4.631-2.13 4.631-4.707v-5.139h-3.692v5.139zm-23.677.94c-.531 0-.939-.4-.939-.94v-5.138H0v5.139c0 2.577 2.13 4.707 4.708 4.707h5.138V25.77H4.631z"
            />
          </svg>
        </button>
      </div>
      <div
        class="react-flow__panel react-flow__attribution bottom right"
        data-message="Please only hide this attribution when you are subscribed to React Flow Pro: https://reactflow.dev/pro"
        style="pointer-events: all;"
      >
        <a
          aria-label="React Flow attribution"
          href="https://reactflow.dev"
          rel="noopener noreferrer"
          target="_blank"
        >
          React Flow
        </a>
      </div>
      <div
        id="react-flow__node-desc-1"
        style="display: none;"
      >
        Press enter or space to select a node.
        You can then use the arrow keys to move the node around.
         Press delete to remove it and escape to cancel.
         
      </div>
      <div
        id="react-flow__edge-desc-1"
        style="display: none;"
      >
        Press enter or space to select an edge. You can then press delete to remove it or escape to cancel.
      </div>
      <div
        aria-atomic="true"
        aria-live="assertive"
        id="react-flow__aria-live-1"
        style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(100%);"
      />
    </div>
  </div>
</div>
`;
